### PR TITLE
Source build plugin

### PR DESCRIPTION
**Pull request by MartinBasti**
Build plugin for building source containers

Maintainers will complete the following section:
- [x] Commit messages are descriptive enough
- [x] "Signed-off-by:" line is present in each commit
- [x] Code coverage from testing does not decrease and new code is covered
- [x] JSON/YAML configuration changes are updated in the relevant schema
- [x] Changes to metadata also update the documentation for the metadata
- (n/a) Pull request includes link to an osbs-docs PR for user documentation updates
  - will be documented in a future change
- [x] New feature can be disabled from a configuration file

### CODE REVIEW

File: atomic_reactor/core.py [hunk 1]
```diff
@@ -379,6 +379,9 @@ def import_image_from_stream(self, *args, **kwargs):
     def get_archive(self, *args, **kwargs):
         return self.tasker.get_archive(*args, **kwargs)
 
+    def cleanup_containers(self, *args, **kwargs):
+        return self.tasker.clenup_containers(*args, **kwargs)
```
Comment (twaugh) [inline atomic_reactor/core.py]: Typo, clenup → cleanup

File: atomic_reactor/plugins/build_sources.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+        if response != 0:
+            reason = (
+                "Source container build failed with error code {}. "
+                "See build logs for details".format(response)
+            )
+            return BuildResult(logs=output, fail_reason=reason)
+
+        # TODO find out how to return path to OCI image
+        # TODO can it be injected into workflow or extend BuildResult?
```
Comment (MartinBasti) [inline atomic_reactor/plugins/build_sources.py]: @twaugh @rcerven @athos-ribeiro 

Proposal:
extend `BuildResult` with `oci_path` attribute which will be mutually exclusive with `image_id`

WDYT?

**Comment by MartinBasti** [comment]
Fix for tests failures https://github.com/containerbuildsystem/atomic-reactor/pull/1274

File: atomic_reactor/plugins/build_sources.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            image,
+            volume_bindings=volume_bindings,
+            command=command
+        )
+        response = self.tasker.wait(container_id)
+        output = self.tasker.logs(container_id, stream=False)
+
+        self.tasker.cleanup_containers(container_id)
+
+        self.log.debug("source container build output:\n %s", '\n'.join(output))  # TODO remove
```
Comment (MartinBasti) [inline atomic_reactor/plugins/build_sources.py]: :eyes:

File: atomic_reactor/plugins/build_sources.py (hunk 2 above)
Comment (twaugh) [reply atomic_reactor/plugins/build_sources.py]: Yes, looks like the right way.

File: atomic_reactor/build.py [hunk 4]
```diff
@@ -75,18 +76,24 @@ def __init__(self, logs=None, fail_reason=None, image_id=None,
...
         assert fail_reason is None or bool(fail_reason), \
             "If fail_reason provided, can't be falsy"
         # must provide one, not both
-        assert bool(fail_reason) != bool(image_id), \
+        assert not (fail_reason and image_id), \
             "Either fail_reason or image_id should be provided, not both"
+        assert not (fail_reason and oci_image_path), \
+            "Either fail_reason or oci_image_path should be provided, not both"
+        assert not (image_id and oci_image_path), \
+            "Either image_id or oci_image_path should be provided, not both"
```
Comment (twaugh) [inline atomic_reactor/build.py]: Starting to wonder whether there should be an inherited `BuildFailureResult`, `BuildImageIdResult`, `BuildOciPathResult`. It's probably fine as it is, just wonder if you considered it?

File: atomic_reactor/constants.py [hunk 5]
```diff
@@ -108,6 +108,8 @@
 PLUGIN_VERIFY_MEDIA_KEY = 'verify_media'
 PLUGIN_EXPORT_OPERATOR_MANIFESTS_KEY = 'export_operator_manifests'
 PLUGIN_PUSH_OPERATOR_MANIFESTS_KEY = 'push_operator_manifests'
+PLUGIN_CONTAINER_SOURCE_BUILD_KEY = 'source_container_build'
```
Comment (twaugh) [inline atomic_reactor/constants.py]: Seeing this alongside the other names I onder if it should be 'build_source_container' (verb at the front).

File: atomic_reactor/plugins/build_sources.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            srpms_path=srpms_path,
+            output_path=output_path,
+        )
+        container_id = self.tasker.run(
+            image,
+            volume_bindings=volume_bindings,
+            command=command
+        )
+        response = self.tasker.wait(container_id)
+        output = self.tasker.logs(container_id, stream=False)
```
Comment (twaugh) [inline atomic_reactor/plugins/build_sources.py]: Is it possible to stream it? No problem if not.

File: atomic_reactor/schemas/config.json [hunk 7]
```diff
@@ -607,6 +607,17 @@
         "items": {
             "type": "string"
         }
+    },
+    "source_containers": {
+        "description": "Configuration options for building source container images",
+        "type": "object",
+        "properties": {
+            "source_image_builder": {
```
Comment (twaugh) [inline atomic_reactor/schemas/config.json]: Wonder if 'image' should be the last word in this name, something like source_builder_image.

File: tests/plugins/test_build_sources.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+from atomic_reactor.plugins.pre_reactor_config import (
+    WORKSPACE_CONF_KEY,
+    ReactorConfig,
+    ReactorConfigPlugin,
+)
+from tests.docker_mock import mock_docker
+from tests.constants import TEST_IMAGE, MOCK_SOURCE
+
+
+SOURCE_CONTAINERS_CONF = {'source_image_builder': "quay.io/vbatts/bsi:latest"}
```
Comment (twaugh) [inline tests/plugins/test_build_sources.py]: quay.io/ctrs/bsi:latest

File: docs/plugins.md [hunk 9]
```diff
@@ -154,6 +154,10 @@ These are run after we have everything ready for build
    * Status: not yet enabled
    * Builds image in remote environment
 
+ * **source_build**
```
Comment (twaugh) [inline docs/plugins.md]: Let's make sure this is consistent with the plugin name.

File: atomic_reactor/plugins/build_sources.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+from atomic_reactor.plugins.pre_reactor_config import get_value
+
+
+class SourceContainerBuilderPlugin(BuildStepPlugin):
+    """
+    Build source container image using
+    https://github.com/containers/BuildSourceImage
+
+    Image https://quay.io/repository/ctrs/bsi should be pushed to image stream
+    on OCP instance and should be used in `source_image_builder`
```
Comment (twaugh) [inline atomic_reactor/plugins/build_sources.py]: Let's make sure this is consistent with the plugin name.

File: atomic_reactor/build.py (hunk 4 above)
Comment (MartinBasti) [reply atomic_reactor/build.py]: Sounds good, however I'd prefer to do it in separate PR and change it consistently across all build plugins

File: atomic_reactor/plugins/build_sources.py (hunk 6 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/build_sources.py]: I don't know yet, for now there is no issue as that log is basically not used. Removed, is better to use generators, we will see when this plugin will be integrated with others

File: atomic_reactor/plugins/build_sources.py (hunk 10 above)
Comment (MartinBasti) [reply atomic_reactor/plugins/build_sources.py]: Reworded

**Comment by MartinBasti** [comment]
Thank you for review, I fixed typos in commit messages

File: atomic_reactor/constants.py [hunk 11]
```diff
@@ -108,6 +108,8 @@
 PLUGIN_VERIFY_MEDIA_KEY = 'verify_media'
 PLUGIN_EXPORT_OPERATOR_MANIFESTS_KEY = 'export_operator_manifests'
 PLUGIN_PUSH_OPERATOR_MANIFESTS_KEY = 'push_operator_manifests'
+PLUGIN_BUILD_SOURCE_CONTAINER_KEY = 'build_source_container'
```
Comment (twaugh) [inline atomic_reactor/constants.py]: Looking at this again, I don't think we should have the 'build' prefix in the plugin name. Comparing with orchestrate_build:
* it's a build plugin
* the plugin's name is orchestrate_build (line 101)
* the file is named build_orchestrate_build.py (${plugintype}_${pluginname}.py)
* the class implementing it is called OrchestrateBuildPlugin

Following that pattern we'd want:
* the plugin's name is source_container
* the file is named build_source_container.py
* the class would be named SourceContainerPlugin (or whatever, this part isn't so important)

Comment (MartinBasti) [reply atomic_reactor/constants.py]: updated, thank you

**Comment by twaugh** [comment]
Shall we document the new config section in osbs-docs?

**Comment by MartinBasti** [comment]
I'd wait until this feature is complete and document it at once

File: atomic_reactor/build.py [hunk 12]
```diff
@@ -75,18 +76,24 @@ def __init__(self, logs=None, fail_reason=None, image_id=None,
                        should be set as labels on OpenShift build
         :param skip_layer_squash: boolean, direct post-build plugins not
                                   to squash image layers for this build
+        :param oci_image_path: str, path to OCI image directory
         """
         assert fail_reason is None or bool(fail_reason), \
             "If fail_reason provided, can't be falsy"
         # must provide one, not both
-        assert bool(fail_reason) != bool(image_id), \
+        assert not (fail_reason and image_id), \
```
Comment (rcerven) [inline atomic_reactor/build.py]: assert not (fail_reason and (image_id or image_id)

instead having 2 checks?

Comment (MartinBasti) [reply atomic_reactor/build.py]: I wanted to have different error msgs

**Comment by MartinBasti** [comment]
not adding release notes as this plugin is not enabled yet and require additional counterparts

**Pull request merged**